### PR TITLE
Add contract Deploy/Update/Destory event

### PR DESCRIPTION
**Pull request by Tommo-L**

### CODE REVIEW

**Review by shargon — COMMENTED** [review]
For me it's good, but the users can already do it

**Comment by devhawk** [comment]
I'm a fan of this change. Can imagine building a contract tracker on top that can provide install/update/delete history like the Nep17 tracker provides transfer history

**Review by devhawk — DISMISSED** [review]

File: src/neo/SmartContract/Native/ManagementContract.cs [hunk 1]
```diff
@@ -21,6 +21,56 @@ public sealed class ManagementContract : NativeContract
...
+            var events = new List<ContractEventDescriptor>(Manifest.Abi.Events)
+            {
+                new ContractEventDescriptor
+                {
+                    Name = "Deploy",
+                    Parameters = new ContractParameterDefinition[]
+                    {
+                        new ContractParameterDefinition()
+                        {
+                            Name = "Contract",
```
Comment (erikzhang) [inline src/neo/SmartContract/Native/ManagementContract.cs:35]: `Hash`?

File: src/neo/SmartContract/Native/ManagementContract.cs [hunk 2]
```diff
@@ -21,6 +21,56 @@ public sealed class ManagementContract : NativeContract
...
+                    }
+                },
+                new ContractEventDescriptor
+                {
+                    Name = "Update",
+                    Parameters = new ContractParameterDefinition[]
+                    {
+                        new ContractParameterDefinition()
+                        {
+                            Name = "Contract",
```
Comment (erikzhang) [inline src/neo/SmartContract/Native/ManagementContract.cs:47]: `Hash`?

File: src/neo/SmartContract/Native/ManagementContract.cs [hunk 3]
```diff
@@ -21,6 +21,56 @@ public sealed class ManagementContract : NativeContract
...
+                            Type = ContractParameterType.Hash160
+                        },
+                        new ContractParameterDefinition()
+                        {
+                            Name = "UpdateCounter",
+                            Type = ContractParameterType.Integer
+                        }
```
Comment (erikzhang) [inline src/neo/SmartContract/Native/ManagementContract.cs:54]: I think this parameter is useless.

File: src/neo/SmartContract/Native/ManagementContract.cs [hunk 4]
```diff
@@ -21,6 +21,56 @@ public sealed class ManagementContract : NativeContract
...
+                    }
+                },
+                new ContractEventDescriptor
+                {
+                    Name = "Destory",
+                    Parameters = new ContractParameterDefinition[]
+                    {
+                        new ContractParameterDefinition()
+                        {
+                            Name = "Contract",
```
Comment (erikzhang) [inline src/neo/SmartContract/Native/ManagementContract.cs:64]: `Hash`?

File: src/neo/SmartContract/Native/ManagementContract.cs [hunk 5]
```diff
@@ -116,6 +166,8 @@ private ContractState Deploy(ApplicationEngine engine, byte[] nefFile, byte[] ma
             if (md != null)
                 engine.CallFromNativeContract(Hash, hash, md.Name, false);
 
+            engine.SendNotification(Hash, "Deploy", new VM.Types.Array { contract.Hash.ToArray() });
```
Comment (erikzhang) [inline src/neo/SmartContract/Native/ManagementContract.cs:169]: You must add `CallFlags.AllowNotify` to the method.

**Review by erikzhang — CHANGES_REQUESTED** [review]

File: src/neo/SmartContract/Native/ManagementContract.cs (hunk 5 above)
Comment (Tommo-L) [reply src/neo/SmartContract/Native/ManagementContract.cs:164]: Done

**Review by erikzhang — APPROVED** [review]

**Review by shargon — APPROVED** [review]

**Pull request merged**